### PR TITLE
Optimize updateAllUserStatus Query

### PR DESCRIPTION
**Pull request by Soumava-221B**
Date: 16/06/2026

Developer Name: Soumava Das

---

## Issue Ticket Number
#2597 

## Description

Optimize the updateAllUserStatus cron job by delegating scheduled status filtering to Firestore to reduce document reads.

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [x] Yes
- [ ] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

</details>

## Test Coverage
<details>
<summary>Screenshot 1</summary>

Screenshot 2026-06-16 at 8 41 52 PM

</details>

### CODE REVIEW

File: models/userStatus.js
Comment (MayankBansal12) [inline models/userStatus.js:1]: please add tests for your change

**Comment by MayankBansal12** [comment]
@Soumava-221B please add proof for your changes, screen recording for the api request
and improve PR title and simplify the description (refer to other merged PRs)

File: test/integration/userStatus.test.js [hunk 1]
```diff
@@ -148,6 +156,114 @@ describe("UserStatus", function () {
       const response = await chai.request(app).patch("/users/status/update").set("cookie", `${cookieName}=${jwt}`);
       expect(response).to.have.status(401);
     });
+
+    it("Should update user status and return 200 for a superuser request", async function () {
```
Comment (MayankBansal12) [inline test/integration/userStatus.test.js]: where's the test for checking that future status documents aren't being read? And wasn't the update behaviour being tested in existing tests?

File: test/integration/userStatus.test.js [hunk 2]
```diff
@@ -148,6 +156,114 @@ describe("UserStatus", function () {
...
+        toFake: ["Date"],
+      });
+      const today = Date.now();
+
+      const userToUpdateId = await addUser(userData[1]);
+      const userNotToUpdateId = await addUser(userData[2]);
+      const userBoundaryUpdateId = await addUser(userData[3]);
+
+      const userToUpdateStatusRef = firestore.collection("usersStatus").doc();
+      await userToUpdateStatusRef.set({
```
Comment (MayankBansal12) [inline test/integration/userStatus.test.js]: the mock value for userStatus can be defined in constants once and reused...there might be existing value that can be reused here

File: test/unit/models/userStatus.js [hunk 3]
```diff
@@ -86,4 +86,141 @@ describe("tasks", function () {
     expect(response.userStatusExists).to.equal(true);
     expect(response.data.futureStatus.state).to.equal("UPCOMING");
   });
+
+  describe("updateAllUserStatus", function () {
```
Comment (MayankBansal12) [inline test/unit/models/userStatus.js:99]: - you can refactor all tests in this describe block, all of them users same userStatus mock data 
- also please remove comments if aren't necessary

**Comment by MayankBansal12** [comment]
@Soumava-221B can you please add proof for your changes? screenshot or any other relevant recording...

**Comment by Soumava-221B** [comment]
> @Soumava-221B can you please add proof for your changes? screenshot or any other relevant recording...

Sure

## Before the cron job run and calls the PATCH API (Query operation benchmark)

Screenshot 2026-07-09 at 7 07 32 PM

## Manually hitting PATCH API endpoint
Screenshot 2026-07-09 at 1 01 21 AM

## After the userStatus is updated (Query operation benchmark)
Screenshot 2026-07-09 at 1 03 45 AM

**Comment by Soumava-221B** [comment]
## Test output of userStatus.test.js (Integration)

Screenshot 2026-07-15 at 9 44 14 AM

## Test output of userStatus.js (Model)
Screenshot 2026-07-15 at 9 44 54 AM

File: test/integration/userStatus.test.js (hunk 1 above)
Comment (Soumava-221B) [reply test/integration/userStatus.test.js]: Test for checking the future status documents is done by the combination of integration and unit test

### Integration Test
In `test/integration/userStatus.test.js`:
It seeds three user status documents:
- `userToUpdateId` (scheduled for yesterday, so it should update).
- `userBoundaryUpdateId` (scheduled for today, so it should update).
- `userNotToUpdateId` (scheduled for tomorrow, so it should NOT update).
It verifies that the document with the future date is ignored during the update run using these assertions:
#### API Response Assertions:
```javascript
expect(response.body.data.usersCount).to.equal(2); // Only matches the 2 eligible documents
expect(response.body.data.oooUsersAltered).to.equal(2);
```

#### Database State Assertion:
```javascript
const nonUpdatedDoc = await userNotToUpdateStatusRef.get();
expect(nonUpdatedDoc.data().currentStatus.state).to.equal(userState.OOO);
expect(nonUpdatedDoc.data().futureStatus.state).to.equal(userState.ACTIVE); // remains untouched
```

### Unit Test
In `test/unit/models/userStatus.js`:
The specific test is: `Should not update user status when futureStatus.from > today (e.g. from is in the future)`.
It seeds a single user status document with a future date (today + 24 hours) and verifies:
```javascript
const summary = await updateAllUserStatus();
expect(summary.usersCount).to.equal(0); // Verifies the query returned 0 documents to be read/updated
expect(summary.oooUsersAltered).to.equal(0);
```

## Previously the bulk update behaviour was not tested in the existing integration tests.

File: test/integration/userStatus.test.js (hunk 2 above)
Comment (Soumava-221B) [reply test/integration/userStatus.test.js]: Currently there is no exsisting constant in the codebase that can be reused directly for these tests.

File: test/integration/userStatus.test.js [hunk 4]
```diff
@@ -148,6 +155,114 @@ describe("UserStatus", function () {
...
+    it("Should update user status and return 200 for a superuser request", async function () {
+      clock = sinon.useFakeTimers({
+        now: new Date("2026-07-14T02:00:00.000Z").getTime(),
+        toFake: ["Date"],
+      });
+      const today = Date.now();
+
+      const userToUpdateId = await addUser(userData[1]);
+      const userNotToUpdateId = await addUser(userData[2]);
+      const userBoundaryUpdateId = await addUser(userData[3]);
```
Comment (MayankBansal12) [inline test/integration/userStatus.test.js]: what's the boundaryUpdateId?

File: test/unit/models/userStatus.js [hunk 5]
```diff
@@ -86,4 +86,117 @@ describe("tasks", function () {
...
+        toFake: ["Date"],
+      });
+    });
+
+    afterEach(async function () {
+      clock.restore();
+      await cleanDb();
+    });
+
+    function buildUserStatusMock(
```
Comment (MayankBansal12) [inline test/unit/models/userStatus.js]: - please use better function naming
- place this func ideally in a utils or helper file so it can be reused in multiple places

File: test/integration/userStatus.test.js [hunk 6]
```diff
@@ -148,6 +155,114 @@ describe("UserStatus", function () {
...
+
+      const userToUpdateStatusRef = firestore.collection("usersStatus").doc();
```
Comment (MayankBansal12) [inline test/integration/userStatus.test.js]: in lines 170, 193, 216, why are you querying collections three times? what behaviour are you testing here? does this match the pattern for logic used in actual function? 


shouldn't the func query one time and then filter accordingly? let me know if i'm missing something

File: test/integration/userStatus.test.js (hunk 2 above)
Comment (MayankBansal12) [reply test/integration/userStatus.test.js]: please define the constant then since you are using this in multiple places or you can reuse the same func you defined for other tests

File: models/userStatus.js [hunk 7]
```diff
@@ -312,32 +316,28 @@ const updateAllUserStatus = async () => {
...
-            nextState: futureState,
-            fallbackTimestamp: today,
-          });
-          if (lastOooUntilUpdate !== undefined) {
-            newStatusData.lastOooUntil = lastOooUntilUpdate;
-          }
-          toUpdate = !toUpdate;
-          summary.oooUsersAltered++;
-        } else {
-          summary.oooUsersUnaltered++;
```
Comment (MayankBansal12) [inline models/userStatus.js:299]: previously there was `oooUsersUnaltered` for keeping track of status that aren't updated, are we still using it anywhere for OOO users who's status aren't changed?

File: models/userStatus.js [hunk 8]
```diff
@@ -297,10 +297,13 @@ const updateAllUserStatus = async () => {
     nonOooUsersUnaltered: 0,
   };
   try {
-    const userStatusDocs = await userStatusModel.where("futureStatus.state", "in", ["ACTIVE", "IDLE", "OOO"]).get();
-    summary.usersCount = userStatusDocs._size;
+    const today = Date.now();
+    const userStatusDocs = await userStatusModel
+      .where("futureStatus.state", "in", ["ACTIVE", "IDLE", "OOO"])
+      .where("futureStatus.from", "<=", today)
+      .get();
```
[the remark this answers is not in the document]
Comment (MayankBansal12) [reply models/userStatus.js:273]: please reply to this with the staging ticket for index creation and mark this as resolved

File: test/integration/userStatus.test.js (hunk 4 above)
Comment (Soumava-221B) [reply test/integration/userStatus.test.js]: The boundaryUpdateId is when the futureStatus.from is set to today

File: models/userStatus.js (hunk 7 above)
Comment (Soumava-221B) [reply models/userStatus.js:299]: Yes, we are still using it check line 296

File: test/integration/userStatus.test.js (hunk 6 above)
Comment (Soumava-221B) [reply test/integration/userStatus.test.js]: Yes, you're right. The function only needs to query once. I'll fix that.

Here's what the three variables represent:

1. `userToUpdate`: The user's OOO period has expired (futureStatus.from is in the past), so the status should transition to `ACTIVE`.
2. `userNotToUpdate`: The user's OOO period is still ongoing (futureStatus.from is in the future), so the status should remain `OOO`.
3. `userBoundaryUpdate`: The OOO period ends exactly today (boundary case), so the status should also transition to `ACTIVE`.

**Comment by Soumava-221B** [comment]
Hi @MayankBansal12 can you review the test files again, i have updated the code

**Comment by Soumava-221B** [comment]
Hi @prakashchoudhary07 can you please review this PR

File: test/integration/userStatus.test.js [hunk 9]
```diff
@@ -192,6 +205,87 @@ describe("UserStatus", function () {
...
+      const today = Date.now();
+      const userToUpdateId = await addUser(userData[1]);
+      const docRef = firestore.collection("usersStatus").doc();
+      await docRef.set(
+        generateOooUserStatusDoc(userToUpdateId, today, {
+          currentStatusFromOffset: -2 * 24 * 60 * 60 * 1000,
+          currentStatusUntilOffset: 2 * 24 * 60 * 60 * 1000,
+          futureStatusFromOffset: -24 * 60 * 60 * 1000,
+        })
+      );
```
Comment (MayankBansal12) [inline test/integration/userStatus.test.js]: @Soumava-221B every test seems to repeat these variables, you can try to refactor and place in before hook

**Comment by MayankBansal12** [comment]
> Hi @MayankBansal12 can you review the test files again, i have updated the code

sorry for the late reply, you can try to refactor the test files by placing repeating declaration in before hook

File: test/integration/userStatus.test.js (hunk 9 above)
Comment (Soumava-221B) [reply test/integration/userStatus.test.js]: I have refactored the test variables

**Comment by Soumava-221B** [comment]
> > Hi @MayankBansal12 can you review the test files again, i have updated the code
> 
> sorry for the late reply, you can try to refactor the test files by placing repeating declaration in before hook

Yes, it's done

**Comment by Soumava-221B** [comment]
Hi @MayankBansal12 gentle reminder, can you review this PR again
